### ISSUE TITLE
Subscribing to multiple stores using HOC

### ISSUE THREAD

**Issue by shivekkhurana** [taviroquai/ReactObservableStore#4]
Hi @taviroquai,

Assume a use case where a component depends on data from 2 stores. I can manually subscribe/unsubscribe to two stores in constructor of that component. 

Can I also extend the package and introduce a function `withStores` (plural) ?
This has a similar footprint like `withStore` except it takes an array of namespaces.

I want to be able to do something like : 
```
import Store from 'react-observable-store';
import { withStores } from 'react-observable-store';

Store.init({ ns1: {foo: 'bar' }, ns2: {bar: 'baz'}});

class MyComponent extends React.Component {};

export default withStores(['ns1', 'ns2'], MyComponent);

```
and access the props like : 

```
this.props.ns1.foo
this.props.ns2.baz
```

I see the `withStore` magic ([here](https://github.com/taviroquai/ReactObservableStore/blob/a339576b888c5ab1fa36ce6162857e8b574884b6/src/ReactObservableStore.js#L126)). And feel that I can simple map over namespaces. Will this break anything else ??

Thanks
PS: Great work with this lib. Have been looking for something to simplify react-state management. This seems to be it. 

Update 0:
With the existing api, I can easily do `withStore('ns1', withStore('ns2', component));` but this will flatten the store state. Something that I do not want.

**Comment by taviroquai** [comment]
@shivekkhurana thanks!

The WrappedComponent merges both passed props with namespace props, having store props overriding passed props IF props have the same name. Is this that you mean when you say "flatten the store state"? Are you using the same prop names in both namespaces?

By design, a component should only subscribe from only one namespace. If we allow to subscribe from multiple namespaces, it can let a user to subscribe from N namespaces (or even all!) which does not seems a good app architecture.

I believe there can be some situations, like a shared component, that has to update when any of the namespaces updates, but for this you can chain the stores has you pointed out. Let's leave it open for now.

**Comment by shivekkhurana** [comment]
Yes. That's exactly what I meant by "flatten the store state". In the example above, the component will have `props == {foo: 'bar', bar: 'baz'}`. 

> Same prop names

I do not have a particular use case right now. I'm just evaluating systems that can replace redux for my apps. The other contender is [@gunn/pure-store](https://github.com/gunn/pure-store). 

> A component should only subscribe from only one namespace

I checked the source and it was evident that it will be hard to subscribe to multiple namespaces. One component subscribing to only one namespace will work for 95% of an average app's use cases. But I've found (over the last 3 years) that, sooner or later, you will be sharing data across namespaces. 

**Follow up question**
I don't mind the chaining of withStore, like the example in update 0. I can easily create a wrapper `withStores` which will make the syntax look a little better.
```
withStores(['ns1', 'ns2'], component);

// instead of 
withStore('ns1', withStore('ns2', component));
```

My only problem is, flattening of namespaces. I tend to use common keys for each domain specific module. For example, if there are two apis: `posts` and `tags`, then both `PostNS` and `TagNS` will have a key called `collection`, representing the list of all posts and tags respectively. 

```
Store.init({
  PostNS: {
    collection: ['Post 1', 'Post 2']
  },
  TagNS: {
    collection: ['Tag 1', 'Tag 2']
  },
})
```

It's very likely that I'd need data from both these namespaces to render the blog list page. If I chain stores, the collection will be overwritten by the top most `withStore` ns. This is my only problem. The easiest solution is to pass props with the namespaces, instead of injecting them directly. 

Instead of accessing the ns keys directly (`this.props.collection`), it's better if I can access them with the ns (`this.props.ns.collection`). This will create clear distinction. IMO this is helpful even if you don't want the ability to subscribe to multiple stores. The distinction makes it clear that these props were injected by ReactOberservableStore. 


How feasible is this ? Does this aligns with your vision ?

**Comment by gregorskii** [comment]
What I ended up doing for this was making a new hoc that supported a map state to props like syntax.

It’s method sig is:

withStore(‘test’, (state) => ({localKey: state}), Comp);

I also handled chanining these with recompose, you can do that and a ton more.

My newer hoc uses the more standard sig:

withStore(‘test’, (state) => ({localKey: state}))(Comp)

As that is supported by recompose ‘compose’:

```
const enhance = compose(
   withStore(‘a’, (state) => ({localKey: state})),
   withStore(‘b’, (state) => ({localKey: state}))
);

enhance(Component)
```

This way the complexity of chaining them together can be an implementation detail in the developers app.

Cheers

**Comment by taviroquai** [comment]
@shivekkhurana Thanks for clarifying and yes the usage of same prop names in different NS can be very common in situations like you pointed out.
I also don't dislike accessing the props in the comp observer under the namespace prop, ie:
```
this.props.ns.foo
```
This was something that I was considering since the beggining and it has its advantagens:
1. no override of props
2. clear distinction that it holds the global store props
... but because I prefer simple solutions over complicating, I ended up just passing the props as it is now.

This would leave wide road to implement observing from multiple namespaces, but of course will break compatibility.

I'm working on a version 2 (branch v2-dev), in which I think there is a very good oportunity to implement this feature.

**Comment by taviroquai** [comment]
@gregorskii Very interesting solution of combining comps.

Did you created your own HOC in your fork?

**Comment by gregorskii** [comment]
Nope its not in this code its project specific.

We did more with our HOC we are using a "container" concept similar to this article:

https://dev.to/washingtonsteven/reacts-new-context-api-and-actions-446o

So we have reducers and actions being fed into the state of the container.

Our container uses the Component version of the Store subscription:

https://github.com/taviroquai/ReactObservableStore#manual-subscribeunsubscribe

**Comment by shivekkhurana** [comment]
@gregorskii 
Interesting use of recompose and enhance. I'm wondering what performance optimisations did you include in your HOC. I know react-redux's connect does some crazy stuff with performance. What's your take on minimising re-renders. Also, is it possible to share that HOC (or maybe parts of it) here ?

@taviroquai Thanks for considering this for `v2-dev`.

**Comment by gregorskii** [comment]
We are managing re-renders in the components, recompose has prebuilt shouldComponentUpdates using onlyRenderForProps.

I can’t share the full container as we built it for a client, but its a pretty standard hoc that simply wraps the component in the “container” that is placed in the first arg. We did not use the withStore version of this in our components because we want consistency between the components that change their state and those that are read only.

We created two things: a render props container component that reads the store in the constructor/ gets updates by subscribing in the cdm and unmounts in cdu. We used similar logic for this container to the article I linked above. Then we created a new hoc that is almost the same as this one, except you pass it one of our container
Components and the component that needs the data, that method signature is above.

We have not gotten to the performance phase, but having a function similar to get state from props helps to control what data the wrapper component seems get, which in turn controls how often they render.

By a rule of thumb we only store one logical object inside of the container, say a container for “widgets” or “todos”.
